### PR TITLE
add H2 Database support - see http://www.h2database.com

### PR DESCRIPTION
**Pull request by gitmotte**
I have inserted a test-dependency and added some classes to support H2 Database (open source - https://www.h2database.com/html/license.html H2 is dual licensed and available under the MPL 2.0 (Mozilla Public License Version 2.0) or under the EPL 1.0 (Eclipse Public License).)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -163,6 +163,12 @@
             <version>1.3</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+	    	<groupId>com.h2database</groupId>
+		    <artifactId>h2</artifactId>
+		    <version>1.4.200</version>
+		    <scope>test</scope>
+		</dependency>
```
Comment (enepomnyaschih) [inline pom.xml:171]: Please unify indentation.

File: src/main/java/com/github/squigglesql/squigglesql/syntax/H2Syntax.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2019 Egor Nepomnyaschih.
```
Comment (enepomnyaschih) [inline src/main/java/com/github/squigglesql/squigglesql/syntax/H2Syntax.java:2]: You may specify yourself as the author of this file.

File: src/main/java/com/github/squigglesql/squigglesql/syntax/H2Syntax.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Override
+  char getIdentifierQuote() {
+    return '"';
+  }
+
+  @Override
+  public void compileEmptyInsert(QueryCompiler compiler) {
+    compiler.write(" () VALUES ()");
+  }
```
Comment (enepomnyaschih) [inline src/main/java/com/github/squigglesql/squigglesql/syntax/H2Syntax.java:30]: Please unify indentation - 4 spaces per step.

File: src/main/java/com/github/squigglesql/squigglesql/syntax/SqlSyntax.java [hunk 4]
```diff
@@ -64,6 +69,8 @@ public static AbstractSqlSyntax from(String protocol) {
                 return MY_SQL_SYNTAX;
             case "postgresql":
                 return POSTGRE_SQL_SYNTAX;
+            case "h2":
+              return H2_SQL_SYNTAX;
```
Comment (enepomnyaschih) [inline src/main/java/com/github/squigglesql/squigglesql/syntax/SqlSyntax.java:73]: Please update JavaDoc for this method.

File: src/test/java/com/github/squigglesql/squigglesql/TestUtils.java [hunk 5]
```diff
@@ -15,15 +15,23 @@
  */
 package com.github.squigglesql.squigglesql;
 
+import java.sql.Connection;
```
Comment (enepomnyaschih) [inline src/test/java/com/github/squigglesql/squigglesql/TestUtils.java:18]: Keep original import ordering. If you use Intellij Idea, you can achieve it by restoring default import settings.

File: src/test/java/com/github/squigglesql/squigglesql/TestUtils.java [hunk 6]
```diff
@@ -35,62 +43,86 @@
  */
 public abstract class TestUtils {
 
-    private static final String USER = "squiggletest";
-    private static final String PASSWORD = "1";
+  private static final Logger logger = Logger.getLogger(TestUtils.class.getName());
+  private static final List<String> BLACKLISTET_URLs = new ArrayList<>();
```
Comment (enepomnyaschih) [inline src/test/java/com/github/squigglesql/squigglesql/TestUtils.java:47]: Please unify indentation - 4 spaces per step.

File: src/test/java/com/github/squigglesql/squigglesql/databases/H2TestDatabase.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Override
+  public String getUrl() {
+    return "jdbc:h2:mem:squiggletest";
+  }
+
+  @Override
+  public void createTable(Connection connection, String name, TestDatabaseColumn[] columns) throws SQLException {
+    try (Statement statement = connection.createStatement()) {
+      statement.execute("CREATE SEQUENCE \"" + name + "_id_seq\"\n" + "START WITH 1\n" + "INCREMENT BY 1\n" + "NO MINVALUE\n" + "NO MAXVALUE\n" + "CACHE 1");
```
Comment (enepomnyaschih) [inline src/test/java/com/github/squigglesql/squigglesql/databases/H2TestDatabase.java:32]: Please avoid lines longer than 120 characters (incl. indentation).

**Review by enepomnyaschih — CHANGES_REQUESTED** [review]
Thank you for your contribution! Please see some remarks. In particular, please follow the same code style as the rest of the project, because, in particular, TestUtils class is now impossible to review.

**Review by enepomnyaschih — CHANGES_REQUESTED** [review]
Please update a list of supported syntaxes in README.md.

**Review by enepomnyaschih — CHANGES_REQUESTED** [review]
In JavaDoc comments to certain methods across the project there are database-specific details like:

* MySQL doesn't support milliseconds. Be careful.
* The corresponding PostgreSQL type is TIMESTAMP WITH TIME ZONE

Please find them and add such details for H2 database for consistency and documentation completeness.

**Comment by gitmotte** [comment]
I'm sorry, my IDE does automatically indent the lines. 

Please fix the formatting yourself, i decided not to use the project because i don't like the syntax. To much boiler-plate code and missing an efficient fluent api. It would be to much work for me to add such a fluent api for the little benefit using sql builders. 

Another missing features are 
* sql JOIN Syntax
* automatic handling of SQL null's in WHERE clauses 
* sql parsing ... 
*

**Comment by enepomnyaschih** [comment]
Sad to hear, but I hope you'll find a tool that better suits you. I want to keep SquiggleSQL free from reflection magic and explicit type casting, so having some boilerplate code is inevitable, yet it makes the library very flexible and transparent. In particular, it helps you completely avoiding code duplication when you build multiple similar SQL queries, which greatly compensates the boilerplate code amount. I agree that the features you listed are indeed important, so we'll likely implement them in the future.

**Comment by enepomnyaschih** [comment]
You don't need to handle SQL null's in WHERE clauses in a special way. If I understand you right, you may just use [IS NOT DISTINCT FROM](https://modern-sql.com/feature/is-distinct-from). If you suggest automatically replacing <value> = NULL with <value> IS NULL, then it won't be a good design decision, because the right operand can end up being NULL as a result of an SQL function call, which is impossible to predict on Java code level, so using IS NOT DISTINCT FROM is the only correct approach to the problem.

**Comment by enepomnyaschih** [comment]
Join syntax has been added and will be included in 4.1.0: https://github.com/squigglesql/squigglesql/blob/development/src/test/java/com/github/squigglesql/squigglesql/TableReferenceInferringTest.java

**Comment by gitmotte** [comment]
> You don't need to handle SQL null's in WHERE clauses in a special way. If I understand you right, you may just use [IS NOT DISTINCT FROM](https://modern-sql.com/feature/is-distinct-from). If you suggest automatically replacing = NULL with IS NULL, then it won't be a good design decision, because the right operand can end up being NULL as a result of an SQL function call, which is impossible to predict on Java code level, so using IS NOT DISTINCT FROM is the only correct approach to the problem.

In my case the right operand is a static value, java null or any Object. But you are right with your remark.

**Comment by gitmotte** [comment]
It would be nice to see a more fluent/builder api. Turn all void methods into methods returning this and add method variants for other methods, the return value may not of interest for all users.

i. e. new SelectQuery(true, a, b).select(c).select(d, e) instead of

SelectQuery q = new SelectQuery(true) ;
q.addToSelection(a);
q.addToSelection(b);
...

Would be nice to see the library splitted into sqlbuilders/parsers and sql Execution (for this purpose i'm using another framework in my project.

Would be nice to see the sqlbuilders independent from sql-domain model, which is not useful for very dynamic queries depending on database-metadata (in my case)

**Comment by enepomnyaschih** [comment]
@gitmotte - I'll think about the fluent API you requested. The reason I've designed it like this was to emphasize that all "add*" methods mutate the query, because a SelectQuery in their return value could lead users to an idea that SelectQuery is immutable and thread-safe, which is wrong. But may be this is just a matter of a clear JavaDoc.

I was also thinking about extracting query execution part to a separate library, but then decided to keep it like this, because it doesn't require any extra dependencies except for built-in java.sql package. Do you feel that SquiggleSQL is too big because of that? Why do you want to split it?

I don't get your idea about sql-domain models. Could you please clarify?

Just curious: if you need an execution-independent SQL builder, do you plan to use prepared statements in order to parametrize your queries, or will your parameters be defined as Literals? In other words, will you choose:

* SELECT product.id FROM product p WHERE p.name = ?
(with ? passed over to JDBC as a parameter)
or
* SELECT product.id FROM product p WHERE p.name = "Apple"

If you use Literals, no further questions, just keep in mind that it doesn't comply with the best practices. But if you use prepared statements, how are you going to achieve it with a different library? I've defined [StatementCompiler](https://github.com/squigglesql/squigglesql/blob/master/src/main/java/com/github/squigglesql/squigglesql/statement/StatementCompiler.java) interface exactly for this purpose - will it be enough to perform this customization?

**Comment by gitmotte** [comment]
> @gitmotte - I'll think about the fluent API you requested. The reason I've designed it like this was to emphasize that all "add*" methods mutate the query, because a SelectQuery in their return value could lead users to an idea that SelectQuery is immutable and thread-safe, which is wrong. But may be this is just a matter of a clear JavaDoc.

yes, i think it just needs clear javadoc.

> I was also thinking about extracting query execution part to a separate library, but then decided to keep it like this, because it doesn't require any extra dependencies except for built-in java.sql package. Do you feel that SquiggleSQL is too big because of that? Why do you want to split it?

it's not the matter of a too big library but a clear separation of concerns. there are many libraries out there which are wrapping simple jdbc - i.e. native jpa , native orm mappers with it's own apis like hibernate, eclipselink and of course i have my own little jdbc-wrapper library for cases i don't like to add those additional bigger dependencies. 

well just to say, that there are many libraries out there which do the same as squiggle-sql but in a more fluent way like querydsl,  jOOQ, ... but all of them depend on a static model generated from database. jpa, jdo additionally add there own query-abstraction-language and include querybuilders. 

> 
> I don't get your idea about sql-domain models. Could you please clarify?
> 
> Just curious: if you need an execution-independent SQL builder, do you plan to use prepared statements in order to parametrize your queries

sure, i'm using prepared statements too, and all of the libraries which allow to execute native queries do that. some of them are really pretty in mapping java-types to sql-set<Type> by themselves or retrieving and converting the values like you need them independent of the conversion-capabilities of the jdbc-driver.

the idea about sql-domain-free sql-builders will be essential, if you don't know much about the model (i.e. building sql-strings from configuration) or need to inspect metadata about the database-model (and build the queries from that metadata) to fulfil your needs. you just have not a fixed model in this case. 

but this needs are out of scope of your library, the purpose is to map a fixed model, build and execute queries against this model.

**Comment by enepomnyaschih** [comment]
Done in 4.1. Merged to "development" branch.

**Comment by enepomnyaschih** [comment]
@gitmotte - Release 4.1.0 has been published at https://search.maven.org/artifact/com.github.squigglesql/squigglesql/4.1.0/jar, and will shortly be promoted to Maven central. See [changelog](https://github.com/squigglesql/squigglesql/blob/master/CHANGELOG.md). I hope that these changes fulfill the majority of your needs. As for the other features you requested:

* I've decided to discard the fluent API proposal for a reason I described before (illusion of immutability) and because addToSelection method already returns ResultColumn value, and therefore it can't be chained. On the other hand, I've substantially simplified Criteria instantiation by providing static methods instead of constructors. For example, instead of

        new MatchCriteria(t.get(ID), MatchCriteria.EQUALS, Parameter.of(id));
    it is now enough to type:

        equal(t.get(ID), Parameter.of(id));
    ...as long as you have static Criteria.* imported
* I'll keep JDBC part separation as a low-priority feature (#14). It is certainly possible, and it will not even change the API too much, but it may require quite a bit of effort, so I'll leave it for the future.
* For sql-domain-free feature, I'm not sure that the design of this feature I'd implement would fulfill your requirements. So please implement such a library according to your vision if you think that it is necessary.

**Pull request closed without merging**